File: src/features/status/entities/repository-status.ts
import { RepositoryFileStatusDto, RepositoryStatusEntryDto } from "../types/repository-status-dto";

export type RepositoryFileStatus =
  | "modified"
  | "added"
  | "deleted"
  | "untracked"
  | "conflicted";

export type RepositoryStatusGroup = "staged" | "unstaged" | "untracked" | "conflicted";

export type RepositoryStatusEntry = {
  path: string;
  status: RepositoryFileStatus;
  staged: boolean;
  group: RepositoryStatusGroup;
};

export function toRepositoryStatusEntry(
  entryDto: RepositoryStatusEntryDto
): RepositoryStatusEntry {
  const status = mapStatus(entryDto.status);
  return {
    path: entryDto.path,
    status,
    staged: entryDto.staged,
    group: resolveGroup(status, entryDto.staged),
  };
}

function mapStatus(statusDto: RepositoryFileStatusDto): RepositoryFileStatus {
  switch (statusDto) {
    case "Modified":
      return "modified";
    case "Added":
      return "added";
    case "Deleted":
      return "deleted";
    case "Untracked":
      return "untracked";
    case "Conflicted":
      return "conflicted";
    default:
      return "modified";
  }
}

function resolveGroup(
  status: RepositoryFileStatus,
  staged: boolean
): RepositoryStatusGroup {
  if (status === "conflicted") {
    return "conflicted";
  }
  if (status === "untracked") {
    return "untracked";
  }
  return staged ? "staged" : "unstaged";
}

File: src-tauri/src/interfaces/dto/repository_status.rs
use serde::{Deserialize, Serialize};

use crate::domain::repository_status::{RepositoryFileStatus, RepositoryStatusEntry};

use super::app_error::AppErrorDto;

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GetRepositoryStatusInputDto {
    pub repository_path: String,
}

#[derive(Debug, Clone, Serialize)]
pub enum RepositoryFileStatusDto {
    Modified,
    Added,
    Deleted,
    Untracked,
    Conflicted,
}

impl From<RepositoryFileStatus> for RepositoryFileStatusDto {
    fn from(value: RepositoryFileStatus) -> Self {
        match value {
            RepositoryFileStatus::Modified => Self::Modified,
            RepositoryFileStatus::Added => Self::Added,
            RepositoryFileStatus::Deleted => Self::Deleted,
            RepositoryFileStatus::Untracked => Self::Untracked,
            RepositoryFileStatus::Conflicted => Self::Conflicted,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RepositoryStatusEntryDto {
    pub path: String,
    pub status: RepositoryFileStatusDto,
    pub staged: bool,
}

impl From<RepositoryStatusEntry> for RepositoryStatusEntryDto {
    fn from(value: RepositoryStatusEntry) -> Self {
        Self {
            path: value.path,
            status: value.status.into(),
            staged: value.staged,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GetRepositoryStatusResponseDto {
    pub entries: Option<Vec<RepositoryStatusEntryDto>>,
    pub error: Option<AppErrorDto>,
}
